rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,513 @@
+Both will be stored under the directory `<userData>/agents/<profile></code>`, e.g.
+`~/Library/Application Support/Teleport Connect/agents/acme.teleport.sh`.
+
+We will not provide the CA pin in the config file. In the MVP, agent connections to the auth server
+will be tunneled through the proxy, so the CA pin is not absolutely necessary. Not using the CA pin
+is also more convenient as we will not have to handle updating the pin after a CA rotation.
+
+The full `teleport node configure` call is:
+
+```
+teleport node configure --output=<userData>/agents/<profile>/config.yaml
+--data-dir=<userData>/agents/<profile>/data
+--proxy=<proxy>
+--token=<join-token>
+```
+
+
+##### Alternatives
+
+In the future, we could consider adding a flag for `teleport start` which makes it read the config
+from standard input, [as suggested by Noah Stride](https://gravitational.slack.com/archives/C0DF0TPMY/p1686663198979029?thread_ts=1686646362.307949&cid=C0DF0TPMY).
+This approach has a number of benefits:
+
+
+* It completely eliminates the need to call `teleport node configure`.
+* It minimizes the API surface of `teleport start` used by Connect, making the interaction between
+  Connect and the agent binary less brittle.
+* It avoids storing the join token on disk.
+* It allows us to easily change the location of the log directory (as noted in the “Log rotation”
+  section).
+* It makes it easier to avoid breaking changes.
+    * If a new agent version requires additional config fields, it’s no issue because the config is
+      recreated from scratch at each launch.
+
+#### Joining the cluster
+
+Once the config file is ready, Connect launches the agent for the first time. The success and
+failure criteria of this step are the same as those of subsequent launches. Both are described in
+the “Launching the agent” section below.
+
+
+#### Ensuring that the setup can be restarted
+
+The user must be able to restart the setup process as each step can independently fail. All steps
+should be prepared to run more than once.
+
+* Setting up the role
+    * Connect must skip calling `UpsertRole` if the role already exists. This is in order to avoid
+      overwriting any changes to the role that the user could have made manually.
+    * Connect must skip calling `UpdateUser` if the user already has the role.
+* Downloading the agent
+    * Connect must skip redownloading the binary if it exists in the sessionData directory.
+* Generating the config file
+    * The join token can be just created again since it will have a short TTL.
+    * `teleport node configure` will fail if the config file already exists, so Connect has to make
+      sure the config file doesn’t exist before calling the command.
+* Joining the cluster
+    * This step doesn’t require any extra accommodations.
+
+### Launching the agent
+
+After the setup is done, Connect launches the agent process with the following command:
```

Comment:
On macOS we get that for free. macOS will perform this check automatically because the agent binary is signed. The archive with binaries is downloaded over HTTPS. I assume `contents.downloadURL` validates the cert since the behavior of this API resembles that of Chrome downloading a file, but this is something worth verifying. 

I don't know if a scenario where a malicious agent replaces the teleport binary with its own version is worth considering – if they're able to do that, they're likely able to do much more damage than that.  

On Linux, I'm not sure what options are available to us as I believe the binaries are not signed in any way. Do we verify the checksum only after download or on every start? Where do we store the checksum?

Assuming that the checksum help us only from running a corrupted binary, the option of ignoring this check altogether seems appealing for the MVP stage.

Reply:
What we haven't tested so far is how [the macOS prompt shown when opening a downloaded program for the first time](https://support.apple.com/en-us/HT202491) affects the process of starting the agent.

I assumed this would be the same as when Connect starts tsh. But this is not guaranteed as tsh is shipped with Connect so the semantics of running a downloaded binary might be different. Does the prompt merely halt the execution of the binary, just like when we start tsh from Connect for the first time, or does it immediately kill the process and attempts to relaunch it after the user clicks "Open"? It's definitely something we have to check.

Reply:
After downloading an archive with our binaries and then attempting to run `teleport` from that archive from the command line, I see the prompt and the execution of the program seems to be paused until I deal with the prompt. There's nothing which suggests that the program is killed.

I also recreated this by running a dev version of Connect with `CONNECT_TSH_BIN_PATH` pointing at freshly extracted tsh.

Overall this means that the setup process is going to work as expected, it's just that it won't be completely hands-off. The user will have to interact with the system dialog.

Reply:
Adding signature or hash validation would be a nice to have, but I don't think we have a significant gap that makes it required.

When considering replacing the binary at a system level I don't think there is much if anything to be gained (since Connect could similarly be modified).

When considering network level attacks, as long as we are doing cert validation, and only have trusted CA's, there is not a significant risk here.  Granted being a user system anything is possible, so some additional validation does have some value.  But I worry that value is limited, and any additional verification may make development more difficult.

Reply:
Given what Mike said, for the MVP we will not be performing any additional validation other than what the given OS provides out of the box.